tests/spiped_nc/graceful_shutdown.c

```diff
@@ -0,0 +1,96 @@
+#include <signal.h>
+#include <stdlib.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+struct graceful_shutdown_cookie {
+	void (* begin_shutdown)(void *);
```

Comment:
As for `cookie`... you'd like to see this?

struct cookie {
        int (* begin_shutdown)(void *);
        void * cookie;
};

Having two different `cookie` names seemed a bit... "off".